TourMate/TourMate/Controllers/PlanPersistenceController.swift

```diff
@@ -5,6 +5,43 @@
 //  Created by Keane Chan on 14/3/22.
 //
 
-struct PlanPersistenceController {
-    // TODO: CRUD for plans - To think of better way to generalize things
+import FirebaseAuth
+
+struct PlanPersistenceController: PlanPersistenceControllerProtocol {
+
+    let firebasePersistenceManager = FirebasePersistenceManager(collectionId: FirebaseConfig.planCollectionId)
+
+    func addPlan(plan: Plan) async -> (Bool, String) {
+        await firebasePersistenceManager.addItem(id: plan.id, item: PlanAdapter.toAdaptedPlan(plan: plan))
+    }
+
+    func fetchPlan(withPlanId planId: String) async -> (Plan, String) {
+        let (adaptedPlan, errorMessage) = await firebasePersistenceManager.fetchItem(id: planId)
+        guard let adaptedPlan = adaptedPlan as? FirebaseAdaptedPlan else {
+            preconditionFailure()
+        }
+
+        let plan = PlanAdapter.toPlan(adaptedPlan: adaptedPlan)
+        return (plan, errorMessage)
+    }
```

Comment:
Maybe something like this?

```suggestion
    func fetchPlan(withPlanId planId: String) async -> (Plan?, String) {
        let (adaptedPlan, errorMessage) = await firebasePersistenceManager.fetchItem(id: planId)
        
        guard let adaptedPlan != nil,
          let adaptedPlan = adaptedPlan as? FirebaseAdaptedPlan,
          let errorMessage.isEmpty else {
                     return (nil, errorMessage)
                   }
        let plan = PlanAdapter.toPlan(adaptedPlan: adaptedPlan)
        return (plan, errorMessage)
    }
```

Reply:
Yup, was thinking something like this